packages/clay-core/src/tree-view/useTree.ts

```diff
@@ -0,0 +1,183 @@
+	extends IExpandable,
+		IMultipleSelection,
+		Pick<ICollectionProps<T>, 'items'> {
+	onItemsChange?: (items: ICollectionProps<T>['items']) => void;
+}
+
+export interface ITreeState<T> extends Pick<ICollectionProps<T>, 'items'> {
+	expandedKeys: Set<Key>;
+	open: (key: Key) => void;
+	reorder: (from: Array<number>, path: Array<number>) => void;
+	selection: IMultipleSelectionState;
+	toggle: (key: Key) => void;
+}
+
+export function useTree<T>(props: ITreeProps<T>): ITreeState<T> {
+	const [expandedKeys, setExpandedKeys] = useInternalState<Set<Key>>({
+		initialValue: new Set(),
+		onChange: props.onExpandedChange,
+		value: props.expandedKeys,
+	});
+
+	const [items, setItems] = useInternalState({
+		initialValue: [],
+		onChange: props.onItemsChange,
+		value: props.items,
+	});
+
+	const selection = useMultipleSelection({
+		onSelectionChange: props.onSelectionChange,
+		selectedKeys: props.selectedKeys,
+	});
+
+	const reorder = (from: Array<number>, path: Array<number>) => {
+		if (!Array.isArray(items)) {
+			return;
+		}
+
+		const tree = createImmutableTree(items);
+
+		tree.produce(from, path);
+
+		setItems(tree.applyPatches());
+	};
+
+	const toggle = (key: Key) => {
+		const expanded = new Set(expandedKeys);
+
+		if (expanded.has(key)) {
+			expanded.delete(key);
+		} else {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
+	const open = (key: Key) => {
+		const expanded = new Set(expandedKeys);
+
+		if (!expanded.has(key)) {
+			expanded.add(key);
+		}
+
+		setExpandedKeys(expanded);
+	};
+
+	return {
+		expandedKeys,
+		items,
+		open,
+		reorder,
+		selection,
+		toggle,
+	};
+}
+
+// RFC 6902 4.4
+type Patch = {
+	op: 'move';
+	from: Array<number>;
+	path: Array<number>;
+};
+
+function createImmutableTree<T extends Array<Record<string, any>>>(tree: T) {
+	const patches: Array<Patch> = [];
+
+	function nodeByPath(path: Array<number>): any {
+		const queue = [...path];
+		const rootIndex: number = queue.shift() as number;
+		let item = tree[rootIndex];
+		let parent = tree;
+		let index = rootIndex;
+
+		while (queue.length) {
+			index = queue.shift() as number;
+
+			if (item.children) {
+				parent = item.children;
+				item = item.children[index];
+			}
+		}
+
+		return {
+			index,
+			item,
+			parent,
+		};
+	}
+
+	function applyPatches(): T {
+		patches.forEach((patch) => {
+			const {from, op, path} = patch;
+
+			switch (op) {
+				// Applies the operation on the tree, the move is functionally
+				// identical to a "remove" operation on the `from` location and
+				// immediately followed by the "add" operation at the target
+				// location with the value that was removed.
+				case 'move': {
+					const nodeToRemove = nodeByPath(from);
+					const parentNode = nodeToRemove.parent.filter(
+						(item: any, index: number) =>
+							index !== nodeToRemove.index
+					);
+
+					tree[from[0]] = {
+						...tree[from[0]],
+						children: parentNode,
+					};
+
+					const {item} = nodeByPath(path);
```

Comment:
@matuzalemsteles I should rename this: it's an item (or node) in the tree, but it's also the "thing" where we going to move the dropped node (maybe we could use `sourceNode` and `destinationNode`). Naming is hard :smile:. Also the bracket notation `...tree[from[0]]` looks a bit ugly. I'll review this part just to make it more readable (if you want to feel free).

Reply:
Quite honestly I liked `nodeByPath` 😂

Reply:
Thanks :smile: , yeah `nodeByPath` is fine, but it's more about that variable `const {item}` that I was talking about. I didn't take time to clean this up today, so I'll continue on monday